crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs

```diff
@@ -66,57 +85,212 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
+    let fix = if let &[edit_expr] = necessary_nodes.as_slice() {
+        // Generate a [`Fix`] for a single type expression, e.g. `int`.
+        Fix::applicable_edit(
+            Edit::range_replacement(checker.generator().expr(edit_expr), annotation.range()),
+            applicability,
+        )
+    } else {
+        match union_type {
+            UnionKind::PEP604 => {
+                generate_pep604_fix(checker, necessary_nodes, annotation, applicability)
+            }
+            UnionKind::TypingUnion => {
+                generate_union_fix(checker, necessary_nodes, annotation, applicability)
+                    .ok()
+                    .unwrap()
+            }
         }
-    } else if has_float && has_int {
-        checker.diagnostics.push(Diagnostic::new(
-            RedundantNumericUnion {
-                redundancy: Redundancy::IntFloat,
-            },
-            annotation.range(),
-        ));
+    };
+
+    checker.diagnostics.push(
+        Diagnostic::new(RedundantNumericUnion { redundancy }, annotation.range()).with_fix(fix),
+    );
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+enum Redundancy {
+    IntFloatComplex,
+    FloatComplex,
+    IntComplex,
+    IntFloat,
+}
+
+impl Redundancy {
+    pub(super) fn from_numeric_flags(numeric_flags: NumericFlags) -> Option<Self> {
+        match numeric_flags.bits() {
+            // NumericFlags::INT | NumericFlags::FLOAT | NumericFlags::COMPLEX
+            0b0111 => Some(Self::IntFloatComplex),
+            // NumericFlags::FLOAT | NumericFlags::COMPLEX
+            0b0110 => Some(Self::FloatComplex),
+            // NumericFlags::INT | NumericFlags::COMPLEX
+            0b0101 => Some(Self::IntComplex),
+            // NumericFlags::INT | NumericFlags::FLOAT
+            0b0011 => Some(Self::IntFloat),
+            _ => None,
+        }
+    }
+}
+
+bitflags! {
+    #[derive(Copy, Clone, Debug, PartialEq, Eq)]
+    pub(super) struct NumericFlags: u8 {
+        /// `int`
+        const INT = 1 << 0;
+        /// `float`
+        const FLOAT = 1 << 1;
+        /// `complex`
+        const COMPLEX = 1 << 2;
+    }
+}
+
+impl NumericFlags {
+    pub(super) fn seen_builtin_type(&mut self, name: &str) {
+        let flag: NumericFlags = match name {
+            "int" => NumericFlags::INT,
+            "float" => NumericFlags::FLOAT,
+            "complex" => NumericFlags::COMPLEX,
+            _ => {
+                return;
+            }
+        };
+        self.insert(flag);
     }
 }
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum UnionKind {
+    /// E.g., `typing.Union[int, str]`
+    TypingUnion,
+    /// E.g., `int | str`
+    PEP604,
+}
+
+// Generate a [`Fix`] for two or more type expressions, e.g. `int | float | complex`.
+fn generate_pep604_fix(
+    checker: &Checker,
+    nodes: Vec<&Expr>,
+    annotation: &Expr,
+    applicability: Applicability,
+) -> Fix {
+    debug_assert!(nodes.len() >= 2, "At least two nodes required");
```

Comment:
The code above handles `nodes.len() < 2` separately and thus this function should never encounter these cases. Is `debug_assert!` ok to denote this?